File: day-2/python/main.py
from sys import argv
from functools import reduce

file_path = argv[1]
red_limit, green_limit, blue_limit = argv[2:5]

limits = {
    "red": int(red_limit),
    "green": int(green_limit),
    "blue": int(blue_limit)
}

def is_game_possible(game_data: list[str]) -> bool:
    for hand in game_data:
        colors = [color.strip().split(' ') for color in hand.split(",")]
        game_failed = any([limits.get(name) < int(count) for count, name in colors])
        if game_failed:
            return False
    return True

def get_cube_power(game_data: list[str]) -> int:
    max_of_color_counts = {
        "red": 0,
        "green": 0,
        "blue": 0
    }
    for hand in game_data:
        colors = [color.strip().split(' ') for color in hand.split(",")]
        for color_count, color_name in colors:
            max_of_color_counts[color_name] = max(max_of_color_counts.get(color_name, 0), int(color_count))

    return reduce(lambda acc, color_score: acc * color_score, max_of_color_counts.values(), 1)
    
possible_games = []

with open(file_path, "r") as file:
    total_cube_power = 0
    for line in file.readlines():
        game_title, game_data = line.split(": ")
        game_data = [hand.strip() for hand in game_data.split(';')]
        total_cube_power += get_cube_power(game_data)
        if is_game_possible(game_data):
            game_number = int(game_title.split(" ")[1])
            possible_games.append(game_number)

result = sum(possible_games)
print(f"Result: {result}")
print(f"Total cube power: {total_cube_power}")
        

File: day-2/typescript/main.ts
import { createReadStream } from "fs";
import { createInterface } from "readline";

const filePath = process.argv[2];
const [red_limit, green_limit, blue_limit] = process.argv.slice(3);

const LIMITS: Record<string, number> = {
  red: Number(red_limit),
  green: Number(green_limit),
  blue: Number(blue_limit),
};

const isGamePossible = (gameData: string[]): boolean => {
  for (const hand of gameData) {
    const colors = hand.split(",").map((color) => color.trim().split(" "));
    const gameFailed = colors.some(
      ([count, name]) => LIMITS[name] < Number(count)
    );
    if (gameFailed) {
      return false;
    }
  }
  return true;
};

const getCubePower = (hands: string[]): number => {
  const maxOfColorCounts: Record<string, number> = {
    red: 0,
    green: 0,
    blue: 0,
  };

  for (const hand of hands) {
    const colors = hand.split(",").map((color) => color.trim().split(" "));
    for (const [count, name] of colors) {
      maxOfColorCounts[name] = Math.max(
        maxOfColorCounts[name] ?? 0,
        Number(count)
      );
    }
  }

  return Object.values(maxOfColorCounts).reduce(
    (acc, colorScore) => acc * colorScore,
    1
  );
};

const possibleGames: number[] = [];
let totalCubePower = 0;

try {
  const readLine = createInterface({
    input: createReadStream(filePath),
    crlfDelay: Infinity,
  });

  readLine.on("line", (line: string) => {
    const [gameName, gameData] = line.split(": ");
    const hands = gameData.split(";").map((hand) => hand.trim());

    totalCubePower += getCubePower(hands);

    if (isGamePossible(hands)) {
      const gameNumber = Number(gameName.split(" ")[1]);
      gameNumber && possibleGames.push(gameNumber);
    }
  });

  readLine.on("close", () => {
    console.log(`Total cube power: ${totalCubePower}`);
    console.log(
      `Possible games score: ${possibleGames.reduce(
        (acc, gameNumber) => acc + gameNumber,
        0
      )}`
    );
  });
} catch (error) {
  console.error(error);
}
